### ISSUE TITLE
Implementing Labels upload functionality

### ISSUE THREAD

**Issue by JesJehle** [Deadwood-ai/deadwood-api#26]
Some questions need to first:
* Upload `label?` as geojson (json) ?
* Upload `aoi` as geojson?
* Upload `label` without aoi possible?
* user sets field `labels_source` and `label_type` by user with drowdown?
* setting `label_quality` by db admin?

**Comment by mmaelicke** [comment]
> Some questions need to first:
> 
> * Upload `label?` as geojson (json) ?

yes. this needs to be a JSON body, where the validation of pydantic is validating a GeoJSON.Geometry<GeoJSON.MultiPolygon> (in Typescript). So not a feature collection or Feature, but only the geometry

> * Upload `aoi` as geojson?

same as above, but a geometry of type `GeoJSON.Polygon`

> * Upload `label` without aoi possible?

No. Not possible. And as far as I understand the data model, labels would be pretty useless without AOI, as  you cannot verify areas where the AI did correctly **not** predict a label, in comparison to areas where the AI was not applied.

> * user sets field `labels_source` and `label_type` by user with drowdown?

You can check the API docs. It's all there: I created enums for this (from the google doc):
* LabelSourceEnum - `visual_interpretation`, `model_prediction`, `fixed_model_prediction` 
* LabelTypeEnum - `point_observation`, `segmentation`, `instance_segmentation`, `semantic_segmentation`

> * setting `label_quality` by db admin?

Currently, this is a mandatory field (all fields are mandatory), so currently the **user** has to provide this. If this is not correct, let me know and I make this field optional.

**Comment by cmosig** [comment]
Regarding the upload on the Website, I would honestly start by allowing simply an file to be uploaded with no special format required because (1) reduce barrier before submission as much as possible and (2) limit implementation effort until we know what we actually need (3) I would be surprised if we get more than three label submissions within the next months.

**Comment by JesJehle** [comment]
The problem is, handling exotic data files like Shapefile in the frontend and the database are problematic. By limiting to geojson (json) we would solve multiple problems:
* its only 1 file
* frontend can understand it
* backend can store it as json in column

**Comment by mmaelicke** [comment]
If we want to support multiple formats, we need different routes anyway. For the frontend, one strict route is needed so that we can provide meaningful error messages. It will not be possible to validate inputs anymore.

Building a route that takes any possible geodata is a major development. We can do that at some point, but it would be a poor choice for the main upload route for labels. We need something to test against, that we know works fine.

Also, I would discourage using shapefiles, as that makes everything just way more complicated.

At the end of the day, it's your call, but accepting any input will cost me a substantial amount of time.

**Comment by cmosig** [comment]
I would not support multiple files, but convert uploaded files manually into the format that your backend requires for now. I expect very rare uploads.

**Comment by mmaelicke** [comment]
Maybe we rather have a meeting about this at some point as we might be talking about different things here: 

Currently we have an upload function that takes GeoJSON geometries and saves them into the database. That is already there. So any other solution requires more implementation work, which is maybe not the most important part if we only expect very rare uploads.
There is no upload for labels to somewhere, so we cannot convert something manually. We would need to implement that from scratch. This upload would be entirely new and would have to take multiple files as needed for a shapefile or aoi and label in different files. 
So, sofar we support geojson(s) uploads and can go for geopackage at some point in the future. Then the discussion would only be if we allow shapefile.
For the 'send us what you have and we do it manually' option I would honestly suggest to let the users mail the stuff to somebody. Building an API for something like seems like a waste of time.

**Comment by cmosig** [comment]
As you find it best. One could also have an upload button during orthophoto submission, where users can upload one zip with all the data. It seems simple to implement but your call. 

An advantage of of managing the label upload through the user interface instead of email is that we already know which file belongs to which ortho. 

Happy to discuss that in a meeting.

**Comment by cmosig** [comment]
I think we need to differentiate between API and Website. I was only talking about the website.

**Comment by mmaelicke** [comment]
Ok, got you. Yeah than we are on the right track. So the process would be 

1. User can upload something (anything) on the website: 
2. if that was a geojson, we can automatically validate and ingest and everone is happy.
3. if that was a geopackage (and possibly shp=, we (later, not yet) read out the layers, validate and ingest
4. if anything else (and 3. for now) we zip that together and either store on the server, or mail someone

**Comment by cmosig** [comment]
Sounds good :+1: 

I would also put a hint on the website that an AOI would is encouraged to submit (but not required) with an info button of what AOI means:

`The area of interest (AOI) is a (Multi)Polygon depicting the labeled area inside an orthophoto. It excludes areas with distortions and includes also non-forested areas.`

**Comment by mmaelicke** [comment]
@JesJehle I guess we move this issue to the frontend-repo now, as it is mainly a frontend issue. As soon as the API needs to provide the addional routes, we can open a new issue here
